codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/transformers/EventStreamNormalizer.kt

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package software.amazon.smithy.rust.codegen.smithy.transformers
+
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.shapes.UnionShape
+import software.amazon.smithy.model.traits.ErrorTrait
+import software.amazon.smithy.model.transform.ModelTransformer
+import software.amazon.smithy.rust.codegen.smithy.traits.SyntheticEventStreamUnionTrait
+import software.amazon.smithy.rust.codegen.util.hasTrait
+import software.amazon.smithy.rust.codegen.util.isEventStream
+
+/**
+ * Generates synthetic unions to replace the modeled unions for Event Stream types.
+ * This allows us to strip out all the error union members once up-front, instead of in each
+ * place that does codegen with the unions.
+ */
+object EventStreamNormalizer {
+    fun transform(model: Model): Model = ModelTransformer.create().mapShapes(model) { shape ->
+        if (shape is UnionShape && shape.isEventStream()) {
+            syntheticEquivalent(model, shape)
+        } else {
+            shape
+        }
+    }
+
+    private fun syntheticEquivalent(model: Model, union: UnionShape): UnionShape {
+        val partitions = union.members().partition { member ->
+            model.expectShape(member.target).hasTrait<ErrorTrait>()
+        }
+        val errorMembers = partitions.first
+        val eventMembers = partitions.second
```

Comment:
nit:
```kotlin
        val (errorMembers, eventMembers) = union.members().partition { member ->
            model.expectShape(member.target).hasTrait<ErrorTrait>()
        }
```

Reply:
Oh nice, I didn't know Kotlin had decomposition syntax like that.